src/gt4py/frontend/gtscript_frontend.py

```diff
@@ -1024,18 +1024,37 @@ def visit_Index(self, node: ast.Index):
         index = self.visit(node.value)
         return index
 
-    def _eval_index(self, node: ast.Subscript) -> Optional[List[int]]:
-        invalid_target = GTScriptSyntaxError(message="Invalid target in assignment.", loc=node)
-
+    def _eval_index(
+        self, node: ast.Subscript, axes: Optional[Set[str]] = None
+    ) -> Optional[List[int]]:
         tuple_or_expr = node.slice.value if isinstance(node.slice, ast.Index) else node.slice
         index_nodes = gtc_utils.listify(
             tuple_or_expr.elts if isinstance(tuple_or_expr, ast.Tuple) else tuple_or_expr
         )
 
         if any(isinstance(cn, ast.Slice) for cn in index_nodes):
-            raise invalid_target
+            raise GTScriptSyntaxError(message="Invalid target in assignment.", loc=node)
         if any(isinstance(cn, ast.Ellipsis) for cn in index_nodes):
             return None
+
+        # Determine if we are using the new-style axis syntax, or the old style.
+        # If this is parsing a data index, this should be fine and will return False.
+        new_style_syntax = False
+        if axes is not None:
+            for index_node in index_nodes:
+                for node in ast.walk(index_node):
+                    if isinstance(node, ast.Name) and node.id in axes:
+                        new_style_syntax = True
+                        break
+
+        if new_style_syntax:
+            index_dict = {axis: 0 for axis in axes}
+            for index_node in index_nodes:
+                axis_context = {axis: gtscript.Axis(axis) for axis in axes}
```

Comment:
I would suggest to change the `ASTEvaluator` implementation to make a copy of the passed context to avoid modifying it, which would also imply the context can be created only once here:

```suggestion
            index_dict = {axis: 0 for axis in axes}
            axis_context = {axis: gtscript.Axis(axis) for axis in axes}
            for index_node in index_nodes:
```